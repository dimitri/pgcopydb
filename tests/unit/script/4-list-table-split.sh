#! /bin/bash

set -x
set -e

# This script expects the following environment variable(s) to be set:
#
#  - PGCOPYDB_SOURCE_PGURI

# `pgcopydb list table-parts` will use the table-size cache populated in
# `pgcopydb.table_size` in '4-list-table-split.sql'.
#
# The values stored in pgcopydb.table_size cache may not accurate, but the
# pgcopydb commands will make decisions based on the available information in
# the cache.

DIR=/tmp/unit/split
OPTS="--not-consistent --split-tables-larger-than 10kB"

pgcopydb list schema --dir ${DIR} ${OPTS} >/dev/null

# Cached size of table_1 is 100 KB, so this will be split into 10 parts
pgcopydb list table-parts --dir ${DIR} \
    --schema-name "public" --table-name "table_1" \
    --split-tables-larger-than "10 kB" 2>&1

# table_2 is identical to table_1 but with the size of 50 KB, so this will be
# split into 5 parts
pgcopydb list table-parts --dir ${DIR} \
    --schema-name "public" --table-name "table_2" \
    --split-tables-larger-than "10 kB" 2>&1

# table_3 doesn't have size in cache, therefore it will not be split
pgcopydb list table-parts --dir ${DIR} \
    --schema-name "public" --table-name "table_3" \
    --split-tables-larger-than "10 kB" 2>&1

# table_ctid_candidate doesn't have a unique integer, therefore it will be split by ctid
pgcopydb list table-parts --dir ${DIR} \
    --schema-name "public" --table-name "table_ctid_candidate" \
    --split-tables-larger-than "10 kB" 2>&1

# ctid split is disabled, no partitioning will be done
pgcopydb list table-parts --dir ${DIR} \
    --schema-name "public" --table-name "table_ctid_candidate_skip" \
    --split-tables-larger-than "10 kB" \
    --skip-split-by-ctid 2>&1

