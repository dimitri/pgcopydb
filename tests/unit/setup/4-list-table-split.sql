---
--- This file creates tables and populate the pgcopdyb.table_size cache with
--  fake sizes.
---

-- Create three tables with identical schema and data

create table table_1 (
    c_bigserial bigserial primary key,
    c_char char(580)
);

create table table_2 (
    c_bigserial bigserial primary key,
    c_char char(150)
);

create table table_3 (
    c_bigserial bigserial primary key,
    c_char char(512)
);

-- Insert 100 rows into table_1 and duplicate data in table_2, table_3 is empty.

insert into table_1 (c_char)
select
    left (md5(random()::text),
        10)
from
    generate_series(1, 100) s (i);

insert into table_2
select
    *
from
    table_1;

--
-- also create tables with names that needs double-quoting to see that our
-- partitioning queries can cope with that
--
CREATE SCHEMA IF NOT EXISTS "Sp1eCial .Char";

CREATE TABLE "Sp1eCial .Char"."source1testing"
 (
   "s0" int PRIMARY KEY,
   "s1" int NOT NULL
 );

insert into "Sp1eCial .Char"."source1testing"("s0", "s1")
select x, (x * 2) % 100000
  from generate_series(1, 10000) AS t(x);


CREATE TABLE "Sp1eCial .Char"."Tabl e.1testing"
 (
  "iD" int PRIMARY KEY,
  "regId" int,
  "status" int,
  "nA M.e" character varying(20) NOT NULL,

   CONSTRAINT "Tabl e_fk_1_testing"
  FOREIGN KEY ("iD")
   REFERENCES "Sp1eCial .Char"."source1testing"("s0")
);

insert into "Sp1eCial .Char"."Tabl e.1testing"("iD", "regId", "status", "nA M.e")
select
    "s0",
    "s0",
    random() * 100,
    'Name ' || "s0"
from
    "Sp1eCial .Char"."source1testing";

-- Create two tables with identical schema and data to test ctid split

create table table_ctid_candidate (
    c_char char(512),
    d_char char(512)
);

create table table_ctid_candidate_skip (
    c_char char(512),
    d_char char(512)
);

-- Insert 100 rows into table_ctid_candidate
insert into table_ctid_candidate (c_char, d_char)
select
    left (md5(random()::text), 10),
    left (md5(random()::text), 10)
from
    generate_series(1, 100) s (i);

insert into table_ctid_candidate_skip
select
    *
from
    table_ctid_candidate;

-- insert into pgcopydb.pgcopydb_table_size (oid, bytes)
--      select tname::regclass,
--             10 * 1024 * 1024
--        from (values ('"Sp1eCial .Char"."source1testing"'),
--                     ('"Sp1eCial .Char"."Tabl e.1testing"')) as t(tname);
