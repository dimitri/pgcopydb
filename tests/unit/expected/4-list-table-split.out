2024-04-26 15:30:56.639 87 INFO   main.c:136                Running pgcopydb version 0.15.59.g9a151a6.dirty from "/usr/local/bin/pgcopydb"
2024-04-26 15:30:56.685 87 INFO   copydb.c:105              Using work dir "/tmp/unit/split"
2024-04-26 15:30:56.688 87 INFO   cli_list.c:1294           Table public.table_1 COPY will be split 11-ways
        Part |          Min |          Max |        Count
-------------+--------------+--------------+-------------
        1/11 |           -1 |           -1 |           -1
        2/11 |            1 |           10 |           10
        3/11 |           11 |           20 |           10
        4/11 |           21 |           30 |           10
        5/11 |           31 |           40 |           10
        6/11 |           41 |           50 |           10
        7/11 |           51 |           60 |           10
        8/11 |           61 |           70 |           10
        9/11 |           71 |           80 |           10
       10/11 |           81 |           90 |           10
       11/11 |           91 |           -1 |           -1

2024-04-26 15:30:56.695 94 INFO   main.c:136                Running pgcopydb version 0.15.59.g9a151a6.dirty from "/usr/local/bin/pgcopydb"
2024-04-26 15:30:56.741 94 INFO   copydb.c:105              Using work dir "/tmp/unit/split"
2024-04-26 15:30:56.744 94 INFO   cli_list.c:1294           Table public.table_2 COPY will be split 6-ways
        Part |          Min |          Max |        Count
-------------+--------------+--------------+-------------
         1/6 |           -1 |           -1 |           -1
         2/6 |            1 |           17 |           17
         3/6 |           18 |           34 |           17
         4/6 |           35 |           51 |           17
         5/6 |           52 |           68 |           17
         6/6 |           69 |           -1 |           -1

2024-04-26 15:30:56.753 101 INFO   main.c:136                Running pgcopydb version 0.15.59.g9a151a6.dirty from "/usr/local/bin/pgcopydb"
2024-04-26 15:30:56.812 101 INFO   copydb.c:105              Using work dir "/tmp/unit/split"
2024-04-26 15:30:56.815 101 INFO   cli_list.c:1275           Table public.table_3 (8192 bytes) will not be split
